internal/resourcestatus/resourcekey.go

```diff
@@ -189,54 +135,68 @@ func addState(bd fleet.BundleDeployment, resources map[fleet.ResourceKey][]fleet
 			Name:       modified.Name,
 		}
 		state := fleet.ResourcePerClusterState{
-			State:     "Modified",
-			ClusterID: cluster,
+			State: "Modified",
 		}
 		if modified.Delete {
 			state.State = "Orphaned"
 		} else if modified.Create {
 			state.State = "Missing"
 		} else if len(modified.Patch) > 0 {
 			state.Patch = &fleet.GenericMap{}
-			err := json.Unmarshal([]byte(modified.Patch), state.Patch)
-			if err != nil {
+			if err := json.Unmarshal([]byte(modified.Patch), state.Patch); err != nil {
 				errors = append(errors, err)
 			}
 		}
-		appendState(resources, key, state)
+		resources[key] = state
 	}
-	return incomplete, errors
+
+	return resources, errors
 }
 
-func appendState(states map[fleet.ResourceKey][]fleet.ResourcePerClusterState, key fleet.ResourceKey, state fleet.ResourcePerClusterState) {
-	if existing, ok := states[key]; ok || key.Namespace != "" {
-		states[key] = append(existing, state)
-		return
-	}
+func aggregateResourceStatesClustersMap(resourceKeyStates resourceStatesByResourceKey) []fleet.Resource {
+	byResourceKey := make(map[fleet.ResourceKey]*fleet.Resource)
+	for resourceKey, entries := range resourceKeyStates {
+		if _, ok := byResourceKey[resourceKey]; !ok {
+			byResourceKey[resourceKey] = &fleet.Resource{
+				Kind:       resourceKey.Kind,
+				APIVersion: resourceKey.APIVersion,
+				Namespace:  resourceKey.Namespace,
+				Name:       resourceKey.Name,
+				State:      "Ready",
+				Type:       toType(resourceKey.APIVersion, resourceKey.Kind),
+				ID:         resourceId(resourceKey.Namespace, resourceKey.Name),
+			}
+		}
+		resource := byResourceKey[resourceKey]
+
+		for _, entry := range entries {
+			if entry.incomplete {
+				resource.IncompleteState = true
+			}
 
-	for k, existing := range states {
-		if k.Name == key.Name &&
-			k.APIVersion == key.APIVersion &&
-			k.Kind == key.Kind {
-			delete(states, key)
-			k.Namespace = ""
-			states[k] = append(existing, state)
+			// "Ready" states are currently omitted
+			if entry.State == "Ready" {
+				continue
+			}
```

Comment:
When I mentioned in the description about `perClusterState` being opinionated, this is one of the causes: `Ready` resources are omitted from this view (and that forces the UI to use `BundleDeployments` for displaying GitRepo resources BTW). If we wanted to change that, it would be as simple as removing these lines.